### PR TITLE
Adding minetest.unregister_item API function

### PR DESCRIPTION
**Pull request by paly2**
Hello, sorry for my bad English,

Sometimes, a mod needs to change another one, and erasing its items may be easier than overriding them.

This is the role of the `minetest.unregister_item` API function. Another function is added by this pull request: `minetest.register_alias_force`. It is used like `minetest.register_alias`, but is able to alias existing items (by deleting them).

For example, [xdecor](https://github.com/minetest-mods/xdecor) and [mobs redo](https://github.com/tenplus1/mobs_redo) both register a cobweb with the same recipe. How to fix this recipe conflict without modifying any mod (and using a third one) (because it's easier for mod updates) ?
1. By clearing a craft using `minetest.clear_craft`
2. By aliasing a cobweb to the other one using `minetest.register_alias_force` as created by this pull request

### CODE REVIEW

File: builtin/game/register.lua [hunk 1]
```diff
@@ -233,13 +256,12 @@ function core.register_alias(name, convert_to)
...
 	end
 	if core.registered_items[name] ~= nil then
-		core.log("warning", "Not registering alias, item with same name" ..
-			" is already defined: " .. name .. " -> " .. convert_to)
-	else
-		--core.log("Registering alias: " .. name .. " -> " .. convert_to)
-		core.registered_aliases[name] = convert_to
-		register_alias_raw(name, convert_to)
+		core.unregister_item(name)
+		core.log("info", "Erased item " .. name .. " because registering an alias with the same name.")
```
Comment (rubenwardy) [inline builtin/game/register.lua]: This change in behaviour may cause problem

Comment (paly2) [inline builtin/game/register.lua]: Would another function like `minetest.register_force_alias` be better ?

Comment (rubenwardy) [inline builtin/game/register.lua]: I'd say so

Comment (paly2) [inline builtin/game/register.lua]: Done !

File: builtin/game/register.lua [hunk 2]
```diff
@@ -172,6 +175,26 @@ function core.register_item(name, itemdef)
...
+		if type == "node" then
+			core.registered_nodes[name] = nil
+		elseif type == "craft" then
+			core.registered_craftitems[name] = nil
+		elseif type == "tool" then
+			core.registered_tools[name] = nil
+		end
+		core.registered_items[name] = nil
+		
+
```
Comment (SmallJoker) [inline builtin/game/register.lua]: 2x whitespace tailing. Please remove these these tabulators.

File: builtin/game/register.lua [hunk 3]
```diff
@@ -172,6 +175,26 @@ function core.register_item(name, itemdef)
 	register_item_raw(itemdef)
 end
 
+function core.unregister_item(name)
+	if not core.registered_items[name] then
+		core.log("warning", "Not unregistering item " .. name .. " because it doesn't exist.")
```
Comment (SmallJoker) [inline builtin/game/register.lua]: This could be done like with the register function: Throw error/warning and exit the function.
```
    return
end
```
It also removes a whole level of indents below.

File: doc/lua_api.txt [hunk 4]
```diff
@@ -212,6 +212,8 @@ Aliases can be added by using `minetest.register_alias(name, convert_to)`.
 This will make Minetest to convert things called name to things called
 `convert_to`.
 
+If an item called name exists, it will be unregistered.
```
Comment (SmallJoker) [inline doc/lua_api.txt]: `name` -> \`name\`

`register_alias` does not unregister an item with `name`. It's what `register_alias_force` does.

File: doc/lua_api.txt [hunk 5]
```diff
@@ -453,6 +455,9 @@ the global `minetest.registered_*` tables.
 * `minetest.register_craftitem(name, item definition)`
     * added to `minetest.registered_items[name]`
 
+* `minetest.unregister_item(name)`
+    * erased from minetest.registered_items[] and from item associated table according to its nature: minetest.registered_nodes[] etc
```
Comment (SmallJoker) [inline doc/lua_api.txt]: Line has > 80 characters. Please break it into two.

File: src/nodedef.cpp [hunk 6]
```diff
@@ -1069,6 +1070,36 @@ content_t CNodeDefManager::allocateDummy(const std::string &name)
...
+	for (std::map<std::string, GroupItems>::iterator i = m_group_to_items.begin();
+		i != m_group_to_items.end() ; ++i) {
+		GroupItems &items = i->second;
+		for (GroupItems::iterator j = items.begin(); j != items.end(); j++) {
+			if (j->first == id) {
+				items.erase(j);
+				j--;
+			}
+		}
+
```
Comment (SmallJoker) [inline src/nodedef.cpp]: Please trim all the tailing whitespaces.

File: src/script/lua_api/l_item.cpp [hunk 7]
```diff
@@ -525,6 +525,29 @@ int ModApiItemMod::l_register_item_raw(lua_State *L)
...
+{
+	NO_MAP_LOCK_REQUIRED;
+	std::string name = luaL_checkstring(L, 1);
+	
+	// Get the writable item definition manager from the server
+	IWritableItemDefManager *idef =
+			getServer(L)->getWritableItemDefManager();
+	
+	// Unregister the node
+	if(idef->get(name).type == ITEM_NODE){
```
Comment (SmallJoker) [inline src/script/lua_api/l_item.cpp]: Please insert spaces after `if` and before `{`
See: [Code style guidelines](http://dev.minetest.net/Code_style_guidelines#Spaces)

Comment (paly2) [inline src/script/lua_api/l_item.cpp]: I tried to respect the style corresponding to the surrounding code.

On line 514 and 518, for example, there is no space after `if` and before `{`.

Comment (SmallJoker) [inline src/script/lua_api/l_item.cpp]: That's generally not a bad idea but these parts do not follow the code style guidelines (yet). Each commit only changes a specific part of the code, thus it will take a while to unify the code style in this project.

**Comment by mgl512** [comment]
> How to fix this recipe conflict without modifying any mod (and using a third one) (because it's easier for mod updates) ?
> 1. By clearing a craft using minetest.clear_craft
> 2. By aliasing a cobweb to the other one using minetest.register_alias_force as created by this pull request

It would be helpful to be able to alias existing nodes. Or the lack of this ability is a pain at least. This feature would provide an easy way to adapt to your needs a few mods slightly overlapping each other.

**Comment by numberZero** [comment]
This will help to solve the old problem with `stairs` (from minetest_game) and `moreblocks`. Currently both register slabs, and `moreblocks` try to alias default ones but fails, so you can find yourself with 2 identically looking but incompatible slab and stair types (at least if installed `moreblocks` in pre-existsing world)

**Comment by paly2** [comment]
@numberZero That's right ! There are many cases where this would be helpful.

**Comment by paly2** [comment]
Is there anything else to fix or change before merging ?

**Comment by nerzhul** [comment]
in fact the real problem is ordering. If xdecor is loaded before this doesn't have any effect

**Comment by Thomas--S** [comment]
@nerzhul IMHO, the ordering problems could maybe be fixed with (soft-)dependencies in depends.txt

**Comment by nerzhul** [comment]
but xdecor should not depend on the overrided mod !

**Comment by Thomas--S** [comment]
If xdecor wants to override nodes from another mod, it should soft-depend on it, IMHO. Isn't it the same as in the `minetest.override_item()` function?

**Comment by t4im** [comment]
I don't think mobs redo has cobwebs. Perhaps you mean another mobs mod, but that isn't important, as xdecor doesn't have anything to do with that one either.

xdecor forked the cobwebs from homedecor.
And these cannot depend on each other; that would be pointless.

**Comment by t4im** [comment]
Mobs redo doesn't have cobwebs. Perhaps you mean another mobs mod, but xdecor still won't have anything to do with that either.

xdecor was forked from homedecor and having those depend on each other, would be kinda pointless.
I assume you will find similar situations, where dependencies simply won't work.

**Comment by paly2** [comment]
Yes, that is also a flaw of the Minetest engine: there is ordering guarantee except using depends. So I thought to use depends, too, as @Thomas--S said (it is already widely used). The overriding mod soft-depends on the overrided one(s).

**Comment by mgl512** [comment]
> in fact the real problem is ordering. If xdecor is loaded before this doesn't have any effect

The idea behind the ability to alias existing nodes from conflicting mods is to do it in your own custom mod. You don't change a line in theirs, you keep them as they are and do upstream updates only. You soft depend on them in your custom mod and do your aliases there.

Add the mod [Fishing](https://forum.minetest.net/viewtopic.php?f=9&t=13659) by Crabman to the 2 above and you have a cobweb hell: 

- mobs:cobweb is in the mod mobs_monsters 
- xdecor defines xdecor:cobweb with the same recipe
- fishing doesn't define any but wants mobs:spider_cobweb (their own fork of mobs) to craft the mithril fishing rod.

It would be very convenient to be able to fix this annoying issues in a custom mod rather than tampering with the code in other people's mods. We need to be able to alias existing nodes to do that.

**Comment by paly2** [comment]
So, because the ordering problem can be easily fixed using depends (and so is not a real problem), why is this PR staying unmerged ?

**Comment by AntumDeluge** [comment]
I need this very much. Thank you. I am building and testing it on my local server. I hope it gets added to the main branch.

**Comment by paly2** [comment]
@AntumDeluge : Thank you for testing and supporting !

File: src/nodedef.h [hunk 8]
```diff
@@ -382,6 +382,8 @@ class IWritableNodeDefManager : public INodeDefManager {
 			const ContentFeatures &def)=0;
 	// If returns CONTENT_IGNORE, could not allocate id
 	virtual content_t allocateDummy(const std::string &name)=0;
+	// Unallocate an id
+	virtual void unallocateName(const std::string &name)=0;
```
Comment (kwolekr) [inline src/nodedef.h]: It's deallocate.  Deallocate, or free.
Not un-allocate.  This is grammatically incorrect English.

File: src/script/lua_api/l_item.cpp [hunk 9]
```diff
@@ -525,6 +525,29 @@ int ModApiItemMod::l_register_item_raw(lua_State *L)
...
+	NO_MAP_LOCK_REQUIRED;
+	std::string name = luaL_checkstring(L, 1);
+
+	// Get the writable item definition manager from the server
+	IWritableItemDefManager *idef =
+			getServer(L)->getWritableItemDefManager();
+
+	// Unregister the node
+	if (idef->get(name).type == ITEM_NODE) {
+		IWritableNodeDefManager *ndef =
```
Comment (kwolekr) [inline src/script/lua_api/l_item.cpp]: Maybe it'd be cleaner to move this assignment outside of the loop

File: src/script/lua_api/l_item.cpp [hunk 10]
```diff
@@ -525,6 +525,29 @@ int ModApiItemMod::l_register_item_raw(lua_State *L)
 	return 0; /* number of results */
 }
 
+// unregister_item(name)
+int ModApiItemMod::l_unregister_item_raw(lua_State *L)
+{
+	NO_MAP_LOCK_REQUIRED;
```
Comment (kwolekr) [inline src/script/lua_api/l_item.cpp]: This function can't be used after the game has started.  This restriction needs to be properly documented and enforced as part of the API.

File: src/script/lua_api/l_item.cpp [hunk 11]
```diff
@@ -525,6 +525,29 @@ int ModApiItemMod::l_register_item_raw(lua_State *L)
...
+	if (idef->get(name).type == ITEM_NODE) {
+		IWritableNodeDefManager *ndef =
+			getServer(L)->getWritableNodeDefManager();
+		ndef->unallocateName(name);
+	}
+
+	// Unregister item
+	idef->unregisterItem(name);
+
+	return 0; /* number of results */
```
Comment (kwolekr) [inline src/script/lua_api/l_item.cpp]: Incorrect comment style.  Please read the [Code Style Guidelines](http://dev.minetest.net/Code_style_guidelines).

Not to mention, that comment does not actually help anybody with understanding the code.  It's unnecessary and too verbose.

File: src/script/lua_api/l_item.cpp [hunk 12]
```diff
@@ -525,6 +525,29 @@ int ModApiItemMod::l_register_item_raw(lua_State *L)
...
+{
+	NO_MAP_LOCK_REQUIRED;
+	std::string name = luaL_checkstring(L, 1);
+
+	// Get the writable item definition manager from the server
+	IWritableItemDefManager *idef =
+			getServer(L)->getWritableItemDefManager();
+
+	// Unregister the node
+	if (idef->get(name).type == ITEM_NODE) {
```
Comment (kwolekr) [inline src/script/lua_api/l_item.cpp]: What happens if this condition is not true?
You want this function to fail silently?

File: src/script/lua_api/l_item.cpp [hunk 13]
```diff
@@ -525,6 +525,29 @@ int ModApiItemMod::l_register_item_raw(lua_State *L)
...
+int ModApiItemMod::l_unregister_item_raw(lua_State *L)
+{
+	NO_MAP_LOCK_REQUIRED;
+	std::string name = luaL_checkstring(L, 1);
+
+	// Get the writable item definition manager from the server
+	IWritableItemDefManager *idef =
+			getServer(L)->getWritableItemDefManager();
+
+	// Unregister the node
```
Comment (kwolekr) [inline src/script/lua_api/l_item.cpp]: These comments are unhelpful.  They just repeat the name of the function being called.  Please remove.

File: src/script/lua_api/l_item.cpp [hunk 14]
```diff
@@ -525,6 +525,29 @@ int ModApiItemMod::l_register_item_raw(lua_State *L)
...
+	if (idef->get(name).type == ITEM_NODE) {
+		IWritableNodeDefManager *ndef =
+			getServer(L)->getWritableNodeDefManager();
+		ndef->unallocateName(name);
+	}
+
+	// Unregister item
+	idef->unregisterItem(name);
```
Comment (kwolekr) [inline src/script/lua_api/l_item.cpp]: What if the item being unregistered didn't exist?  You don't want to report that error to the user?

File: doc/lua_api.txt [hunk 15]
```diff
@@ -453,6 +459,10 @@ the global `minetest.registered_*` tables.
 * `minetest.register_craftitem(name, item definition)`
     * added to `minetest.registered_items[name]`
 
+* `minetest.unregister_item(name)`
+    * erased from minetest.registered_items[] and from item associated table
```
Comment (kwolekr) [inline doc/lua_api.txt]: English please

**Comment by kwolekr** [comment]
I noticed this PR actually adds two API functions: minetest.register_alias_force and minetest.unregister_item.
Perhaps the commit message should be updated to reflect this.

File: src/nodedef.h (hunk 8 above)
Comment (paly2) [inline src/nodedef.h]: OK, I'm not english-native

File: builtin/game/register.lua [hunk 16]
```diff
@@ -242,6 +265,19 @@ function core.register_alias(name, convert_to)
 	end
 end
 
+function core.register_alias_force(name, convert_to)
+	if forbidden_item_names[name] then
+		error("Unable to register alias: Name is forbidden: " .. name)
+	end
+	if core.registered_items[name] ~= nil then
+		core.unregister_item(name)
+		core.log("info", "Erased item " .. name .. " because registering a force alias with the same name.")
```
Comment (kwolekr) [inline builtin/game/register.lua]: This line looks way too long.

File: src/script/lua_api/l_item.cpp (hunk 9 above)
Comment (paly2) [inline src/script/lua_api/l_item.cpp]: I do not think so.

File: src/script/lua_api/l_item.cpp (hunk 10 above)
Comment (paly2) [inline src/script/lua_api/l_item.cpp]: OK

File: src/script/lua_api/l_item.cpp (hunk 11 above)
Comment (paly2) [inline src/script/lua_api/l_item.cpp]: OK, my teacher teached me to comment a lot, but you are write, the function name is clear enough

File: src/script/lua_api/l_item.cpp (hunk 13 above)
Comment (paly2) [inline src/script/lua_api/l_item.cpp]: OK

File: src/nodedef.cpp [hunk 17]
```diff
@@ -1069,6 +1070,36 @@ content_t CNodeDefManager::allocateDummy(const std::string &name)
...
+		// Check if group is empty
```
Comment (kwolekr) [inline src/nodedef.cpp]: Useless comment.  You're repeating **what** the code does.
Comments need to explain **why** the code does what it's doing.

File: src/script/lua_api/l_item.cpp (hunk 14 above)
Comment (paly2) [inline src/script/lua_api/l_item.cpp]: This is done in lua (builtin/game/register.lua).

File: doc/lua_api.txt (hunk 15 above)
Comment (paly2) [inline doc/lua_api.txt]: Hmm please say me what would be the correct sentence, really I am not english-native

File: src/nodedef.cpp [hunk 18]
```diff
@@ -1069,6 +1070,36 @@ content_t CNodeDefManager::allocateDummy(const std::string &name)
...
+	if (m_name_id_mapping.getId(name, id)) {
+		m_name_id_mapping.eraseName(name);
+		m_name_id_mapping_with_aliases.erase(name);
+	}
+
+	// Erase node content from all groups it belongs to
+	for (std::map<std::string, GroupItems>::iterator i = m_group_to_items.begin();
+		i != m_group_to_items.end() ; ++i) {
+		GroupItems &items = i->second;
+		for (GroupItems::iterator j = items.begin(); j != items.end(); j++) {
```
Comment (kwolekr) [inline src/nodedef.cpp]: Could you use less confusing variable names for iterators than i and j?  Something that's not as easy to mix up preferably.

File: src/nodedef.cpp [hunk 19]
```diff
@@ -1069,6 +1070,36 @@ content_t CNodeDefManager::allocateDummy(const std::string &name)
...
+			if (j->first == id) {
+				items.erase(j);
+				j--;
```
Comment (kwolekr) [inline src/nodedef.cpp]: Wrong.  After `items.erase(j)`, the iterator `j` has become invalidated.  Further accesses to this variable are invalid.

File: builtin/game/register.lua (hunk 16 above)
Comment (paly2) [inline builtin/game/register.lua]: Really ? Do you want to make the log message shorter or to break the line ?

File: src/nodedef.cpp [hunk 20]
```diff
@@ -1069,6 +1070,36 @@ content_t CNodeDefManager::allocateDummy(const std::string &name)
...
+			}
+		}
+
+		// Check if group is empty
+		if (items.size() == 0)
+			m_group_to_items.erase(i->first);
```
Comment (kwolekr) [inline src/nodedef.cpp]: This action might invalidate the iterator `i`.

File: src/nodedef.cpp (hunk 17 above)
Comment (paly2) [inline src/nodedef.cpp]: OK, you are write, this comment is useless. However the previous one (`Erase node content from all groups it belongs to`) tells what the code does, but in my opinion it is helpful.

File: src/script/lua_api/l_item.cpp (hunk 9 above)
Comment (kwolekr) [inline src/script/lua_api/l_item.cpp]: I do.

File: src/nodedef.cpp (hunk 18 above)
Comment (paly2) [inline src/nodedef.cpp]: For example, what name (especially for the first one, for the second I think `item` is fine right) ?

File: src/nodedef.cpp (hunk 19 above)
Comment (paly2) [inline src/nodedef.cpp]: OK, I will replace it by a C-like for loop (with an integer iterator)

File: src/nodedef.cpp (hunk 20 above)
Comment (paly2) [inline src/nodedef.cpp]: Why ?

File: builtin/game/register.lua (hunk 16 above)
Comment (kwolekr) [inline builtin/game/register.lua]: The sentence is completely ungrammatical anyway.  Change it to "Removed item " ..name.." while attempting to force add an alias" or similar.

**Comment by paly2** [comment]
@kwolekr: I will do all these changes as soon as I have time.

File: src/nodedef.cpp (hunk 17 above)
Comment (kwolekr) [inline src/nodedef.cpp]: Sure, I agree.  The previous comment is fine, this one is not.
It's "correct" or "right", not "write", by the way.

File: src/nodedef.cpp (hunk 19 above)
Comment (kwolekr) [inline src/nodedef.cpp]: Or you could just remove items using an iterator the right way instead...

File: src/nodedef.cpp (hunk 20 above)
Comment (kwolekr) [inline src/nodedef.cpp]: Because std::map::erase() invalidates any iterators referencing any of the elements erased.
See: http://en.cppreference.com/w/cpp/container/map/erase

File: src/script/lua_api/l_item.cpp (hunk 9 above)
Comment (paly2) [inline src/script/lua_api/l_item.cpp]: Note: are you talking about the condition when you say "the loop" ? I say it again, I am not english native :p

File: src/script/lua_api/l_item.cpp (hunk 14 above)
Comment (kwolekr) [inline src/script/lua_api/l_item.cpp]: Okay, I guess that's fair enough.  Since `unregister_item_raw` is technically undocumented you're able to get away with this.

File: src/nodedef.cpp [hunk 21]
```diff
@@ -1069,6 +1070,36 @@ content_t CNodeDefManager::allocateDummy(const std::string &name)
...
+	// Erase name from name ID mapping
+	content_t id = CONTENT_IGNORE;
+	if (m_name_id_mapping.getId(name, id)) {
+		m_name_id_mapping.eraseName(name);
+		m_name_id_mapping_with_aliases.erase(name);
+	}
+
+	// Erase node content from all groups it belongs to
+	for (std::map<std::string, GroupItems>::iterator i = m_group_to_items.begin();
+		i != m_group_to_items.end() ; ++i) {
```
Comment (kwolekr) [inline src/nodedef.cpp]: This needs to be indented an extra level because it's a control structure continuation.

File: src/nodedef.cpp (hunk 19 above)
Comment (paly2) [inline src/nodedef.cpp]: What is the "right way" to do it ?

File: src/nodedef.cpp (hunk 20 above)
Comment (paly2) [inline src/nodedef.cpp]: Oh, I did not realize that i->first was a reference to the map element, I though it was a copy

File: src/nodedef.cpp (hunk 18 above)
Comment (kwolekr) [inline src/nodedef.cpp]: Maybe something like `iter_groups` instead of `i`, and `iter_groupitems` instead of `j`.
I know this sounds overbearing, but having two iterator variables 'i' and 'j' was the root cause of a tough-to-find bug a while ago - I would rather not give history the chance to repeat itself.

File: src/script/lua_api/l_item.cpp (hunk 14 above)
Comment (paly2) [inline src/script/lua_api/l_item.cpp]: `register_item_raw` and `register_alias_raw` do exactly the same anyway ;-)

File: src/nodedef.cpp (hunk 21 above)
Comment (paly2) [inline src/nodedef.cpp]: OK

File: src/nodedef.cpp (hunk 19 above)
Comment (kwolekr) [inline src/nodedef.cpp]: First result on google:
http://stackoverflow.com/questions/596162/can-you-remove-elements-from-a-stdlist-while-iterating-through-it

File: src/nodedef.cpp (hunk 20 above)
Comment (kwolekr) [inline src/nodedef.cpp]: i->first is the key in the map. Calling std::map::erase on the key will delete any records with 'key' and invalidate iterators referencing it.
If you are iterating through a container and you delete an item within that container at the same time, then the iterator might now be invalidated if it so happens that the iterator is at the same position of the item being erased.

**Comment by AntumDeluge** [comment]
I was going to suggest a method 'unregister_node' as well. But, this already manages that. I assumed that 'unregister_item' only deleted craft items. But that is not the case. Thank you paly2.

File: src/script/lua_api/l_item.cpp (hunk 9 above)
Comment (kwolekr) [inline src/script/lua_api/l_item.cpp]: No, I mean the `IWritableNodeDefManager *ndef = ...`
If you move it outside of the loop then you only call getNodeDefManager() once, and reuse the same nodedef throughout the operation.  Remember, an IGameDef is an abstract interface, and maybe in the future getNodeDefManager might not just return a private member variable.  What if it does a heavier operation?  What if a new IGameDef implementation implements getNodeDefManager in a different quirkier way?

File: src/script/lua_api/l_item.cpp (hunk 11 above)
Comment (kwolekr) [inline src/script/lua_api/l_item.cpp]: "Commenting a lot" isn't really a good thing on its own.  Comments need to have purpose, for example, explaining something unclear to a reader where the purpose of a piece of code might not be immediately clear.  If you're just going to repeat what the code itself says, then there's no value added, and instead adds noise to the code that would otherwise be perfectly clean.

In this particular case, it is generally understood from the surrounding context that `return 0;` is part of the interface between Lua and C++.  If the developer reading this code doesn't know what `return 0` does in this context, perhaps that's a sign that they should read the [Programming in [Lua](http://www.lua.org/pil/contents.html) handbook before writing Minetest code for the Lua scripting interface.

File: doc/lua_api.txt (hunk 15 above)
Comment (kwolekr) [inline doc/lua_api.txt]: Tell me*, you mean.  Not "say me".
When "say" is used with an indirect object such as "me" (i.e. in dative form), you must use the preposition "to" before the object.  "Please say to me" would be grammatical at least, but awkward.  When using "say" imperatively and your sentence has an indirect object, use "tell" instead.

Anyway, to respond to your actual comment:
Why not something like "Unregisters the item `name` from engine, and deletes the entry with key `name` from `minetest.registered_items`."

File: src/script/lua_api/l_item.cpp (hunk 9 above)
Comment (paly2) [inline src/script/lua_api/l_item.cpp]: But there is no loop, just a condition

**Comment by paly2** [comment]
@AntumDeluge: Of course it does ;-)

File: src/script/lua_api/l_item.cpp (hunk 9 above)
Comment (kwolekr) [inline src/script/lua_api/l_item.cpp]: touché.

sorry about that.

File: doc/lua_api.txt (hunk 15 above)
Comment (paly2) [inline doc/lua_api.txt]: But the item is also deleted from `registered_nodes` or `registered_craftitems` or `registered_tools`, and that should be pointed out, right ? That is what I mean in "and from item associated table"

File: src/script/lua_api/l_item.cpp (hunk 11 above)
Comment (paly2) [inline src/script/lua_api/l_item.cpp]: Furthermore, the minetest developers are not in a position to say that my comments are useless (even if there are right), because they wrote by themselves hundred of useless comments that are still there. Just in this file, look at [this comment](https://github.com/minetest/minetest/blob/master/src/script/lua_api/l_item.cpp#L476), [this one](https://github.com/minetest/minetest/blob/master/src/script/lua_api/l_item.cpp#L498), [this one](https://github.com/minetest/minetest/blob/master/src/script/lua_api/l_item.cpp#L510), or [this one](https://github.com/minetest/minetest/blob/master/src/script/lua_api/l_item.cpp#L535). I was based on these comments to write mines.

Comment (rubenwardy) [inline src/script/lua_api/l_item.cpp]: As a general rule, don't look at surrounding code for code style. All of those comments you linked to were written by sapier, who is not known for good code style

File: src/script/lua_api/l_item.cpp (hunk 9 above)
Comment (paly2) [inline src/script/lua_api/l_item.cpp]: No problem :-)

File: src/script/lua_api/l_item.cpp (hunk 10 above)
Comment (kwolekr) [inline src/script/lua_api/l_item.cpp]: Uh oh... which means that all the other register functions have the same issue.
This is a bug that definitely needs to be addressed, but should be done in a separate PR.

File: src/nodedef.cpp [hunk 22]
```diff
@@ -1069,6 +1070,40 @@ content_t CNodeDefManager::allocateDummy(const std::string &name)
...
+	for (std::map<std::string, GroupItems>::iterator iter_groups =
+			m_group_to_items.begin();
+			iter_groups != m_group_to_items.end();) {
+		GroupItems &items = iter_groups->second;
+		for (GroupItems::iterator iter_groupitems = items.begin();
+				iter_groupitems != items.end();) {
+			if (iter_groupitems->first == id)
+				items.erase(iter_groupitems++);
+			else
+				iter_groupitems++;
```
Comment (kwolekr) [inline src/nodedef.cpp]: Unless there's a specific reason for doing so (such as postfix incrementing the argument in an erase call), iterators should be advanced with the prefix increment operator.

Comment (paly2) [inline src/nodedef.cpp]: Why ? The prefix and postfix increment operators have exactly the same effect here (http://en.cppreference.com/w/cpp/concept/ForwardIterator)

Comment (kwolekr) [inline src/nodedef.cpp]: See this:  https://google.github.io/styleguide/cppguide.html#Preincrement_and_Predecrement

Comment (paly2) [inline src/nodedef.cpp]: From your link:
> The tradition developed, in C, of using post-increment when the expression value is not used, especially in for loops.

Very funny! http://dev.minetest.net/Code_style_guidelines#Nested_for_loop_exception => Look at the increment operator.
I should respect this styleguide sheet, but the (wo)man who writted the Minetest Code style Guidelines did not!

File: src/nodedef.cpp [hunk 23]
```diff
@@ -1069,6 +1070,40 @@ content_t CNodeDefManager::allocateDummy(const std::string &name)
...
+		}
+
+		// Check if group is empty
+		if (items.size() == 0)
```
Comment (nerzhul) [inline src/nodedef.cpp]: i think you should increment iter_groups and erase only iter_groups after (this permit to remove else condition)

Comment (paly2) [inline src/nodedef.cpp]: But I would need to copy `iter_groups` first then, which is less convenient than having an else condition

Comment (nerzhul) [inline src/nodedef.cpp]: i don't know why we should copy it.

The fix is to do:

iter_groups++;
if (items.size() == 0) {
m_group_to_items.erase(iter_groups);
}

Comment (paly2) [inline src/nodedef.cpp]: Because the suffix increment operator is applied _after_ the expression is evaluated. In your "fix", the erased group is the next one, and anyway it invalidates the iterator.

File: src/nodedef.cpp (hunk 22 above)
Comment (kwolekr) [inline src/nodedef.cpp]: It's just good practice.  I think your PR is acceptable as-is though.

File: src/nodedef.cpp (hunk 23 above)
Comment (kwolekr) [inline src/nodedef.cpp]: Nerz, no.  Your comment demonstrates a fundamental lack of understanding on the semantics of the increment operator.

paly2:  Do not listen to this comment.  It absolutely will introduce a bug into your code which is fine as-is right now.

**Comment by kwolekr** [comment]
LGTM :+1:

**Comment by Zeno-** [comment]
LGTM :+1:

**Comment by kwolekr** [comment]
https://github.com/minetest/minetest/commit/aa33166386f737f213f1f3005ffd6a6adfd2d97f

**Comment by AntumDeluge** [comment]
Yay!!!
